### ISSUE TITLE
vs 15.3 breaks Installer.2017

### ISSUE THREAD

**Issue by crutkas** [Microsoft/WindowsTemplateStudio#955]
Installed VS 15.3 on a new machine, this error is being thrown.

Assembly 'Microsoft.VisualStudio.Shell.15.0' with identity 'Microsoft.VisualStudio.Shell.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' uses 'Microsoft.VisualStudio.Threading, Version=15.3.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'Microsoft.VisualStudio.Threading' with identity 'Microsoft.VisualStudio.Threading, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'	Installer.2017	C:\Users\crutkas\Source\Repos\WTS\code\src\Installer.2017\CSC	1	Active

**Comment by ralarcon** [comment]
Reopeining as referencing the Microsoft.VisualStudio.Threading 15.3 assembly avoids the extension to continue working in the previous Visual Studio 2017 releases with the following error registered:

![image](https://user-images.githubusercontent.com/2638875/29573161-90883a38-875e-11e7-9b61-f3b7c26f48c9.png)

@crutkas @mrlacey we need to define if we will want to continue supporting previous VS 2017 versions or we are going to force the users to have the latest update.

**Comment by mrlacey** [comment]
I'd argue for the newer version of WTS to target the latest version of VS. If we can enforce this with the installer.

Can we say that to use WTS 1.3 you have to be on VS 15.3?
(If on an older version of VS can keep using the older version of WTS.)

How do other extensions handle issues like this? (@onovotny ?)

**Comment by crutkas** [comment]
I am in agreeement. Come FCU we will need 15.4 as well

Can the extension force 15.3 on update?  We can't compile the extension with 15.3 without this update.

**Comment by mrlacey** [comment]
Can we reference 15.0 & 15.3 assemblies and have it be smart about what's loaded? (Just thinking aloud as I catch up with everything that's been going on)

**Comment by onovotny** [comment]
I'm not sure about how other extensions handle this. In general, I thought that most VS extensions were fixed with the assembly version to avoid this within a major VS version. @AArnott, any suggestions?

**Comment by onovotny** [comment]
Given the servicing mechanism of VS 15, I would have no issue targeting only 15.3/latest stable channel. That said, I imagine someone in Dev Div has metrics on how many ppl are still on 15.0, 15.1, 15.2, etc.

**Comment by gcaughey** [comment]
Agreed. I see VS 15 as evergreen.

**Comment by AArnott** [comment]
Targeting 15.0 is fine and that should work on 15.3. So if you're interested in targeting 15.0 and higher, you should reference MS.VS.Threading 15.0 rather than 15.3. Lots of code and extensions reference 15.0 and run on 15.3 without a problem.

To help resolve the original problem report, I'd need to know more about it, including where the error appears. The best way to target Dev15.0 is to use NuGet packages for the VS SDK assemblies.

**Comment by crutkas** [comment]
So @AArnott your suggestion is to adjust how we reference the VS SDK itself and do it via nuget, not from the built in VS Installer, correct?

**Comment by AArnott** [comment]
Yes. Otherwise you end up targeting your extension to run on whatever version of VS happens to be installed when you build it. But the version you target should be an intentional decision you make.

**Comment by mrlacey** [comment]
We are (were) using NuGet to get the 15.0 version of  MS.VS.Threading

From packages.config.
  `<package id="Microsoft.VisualStudio.Threading" version="15.0.240" targetFramework="net462" />`

Doing so was causing these exceptions

> Assembly 'Microsoft.VisualStudio.Shell.15.0' with identity 'Microsoft.VisualStudio.Shell.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' uses 'Microsoft.VisualStudio.Threading, Version=15.3.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'Microsoft.VisualStudio.Threading' with identity 'Microsoft.VisualStudio.Threading, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

> Cannot find the interop type that matches the embedded interop type 'Microsoft.VisualStudio.Shell.Interop.IVsToolboxItemProvider2'. Are you missing an assembly reference?

To solve the second issue I added the  Microsoft.VisualStudio.Shell.Interop.15.3.DesignTime.dll reference (as suggested at https://developercommunity.visualstudio.com/content/problem/75434/cannot-find-the-interop-type-that-matches-the-embe.html ) and updated the other references to get them in sync.

I think this issue needs someone with a greater understanding of VS extensions and testing version compatibility than I possess.

**Comment by AArnott** [comment]
Are you using NuGet to acquire the `Microsoft.VisualStudio.Shell.15.0.dll` assembly though? It sounds very much like you are getting that assembly from the VS install. That's the problem.

**Comment by mrlacey** [comment]
@AArnott Shell.15.0 is coming from NuGet too

`<package id="Microsoft.VisualStudio.Shell.15.0" version="15.0.26606" targetFramework="net462" />`

It looks like everything is coming from Nuget (based on [packages.config](https://github.com/Microsoft/WindowsTemplateStudio/blob/dev/code/src/Installer.2017/packages.config) and [csproj](https://github.com/Microsoft/WindowsTemplateStudio/blob/dev/code/src/Installer.2017/Installer.2017.csproj))
Any ideas?

**Comment by crutkas** [comment]
taking the nuclear option, i kicked off an email with Unni.

**Comment by mrlacey** [comment]
reassigning to @ralarcon as he said on email he'd look at it and he'll have a chance before I do

**Comment by AArnott** [comment]
```xml
<package id="Microsoft.VisualStudio.Shell.15.0" version="15.0.26606" targetFramework="net462" />
```

That version is actually from 15.3. That's your problem. You should use the 15.0.26201 version of that package.

It's a poor package versioning scheme. One that we hope to improve next time.

**Comment by ralarcon** [comment]
@AArnott  The production version was referencing the nuget packages for version 15.0.26201. I guess @mrlacey get that line from dev branch (which was updated but have the problems of being executed in previous versions of Vs2017 lower than 15.3).

Well, I have cleaned up all references from all projects using Microsoft.VisualStudio.Shell.15.0 (v15.0.26606) and Microsoft.VisualStudio.Threading v15.3. Then a have re-added the references to Microsoft.VisualStudio.15.0 (v15.0.26201) and Microsoft.VisualStudio.Threading v15.0.240. Now the solution builds in Visual Studio 2017 update 3 (15.3). Some cleanup magic I would say.

Here you can check that the master branch projects where referencing Microsoft.VisualStudio.Shell.15.0 thru nuget to the v15.0.26201 
https://github.com/Microsoft/WindowsTemplateStudio/blob/e4642ed6af1ec6e32249d12e26de492159d231e1/code/src/Installer.2017/packages.config#L13

I took the chance to remove some references to Microsoft.VisualStudio.Shell.14.0 which looks to be not required anymore.

In summary, removing all references and re-adding those, made the solution to build properly again in Visual Sutido 2017 15.3. Probably there was a reference to the SDK instead to the nuget somewhere, but I didn't be able to find it.

Finall step pending is to check that the extension now runs fine Visual Studio 15.2 or lower.

**Comment by ralarcon** [comment]
Closed by error.

**Comment by ralarcon** [comment]
Finally I built the extension in VS2017 15.3 and tested in a vanilla machine with VS2017 15.0 and worked fine. Upgraded the testing machine to 15.3 and works fine. With this we fulfill the backward compatibilty requirements.

Changes are merged in #1007

Thanks everyone for his help.

**Comment by crutkas** [comment]
woot!

**Comment by crutkas** [comment]
can compile now
